cli/f3.go

```diff
@@ -39,25 +46,161 @@ var (
+							if cctx.Args().Present() {
+								var instance uint64
+								instance, err = strconv.ParseUint(cctx.Args().First(), 10, 64)
+								if err != nil {
+									return fmt.Errorf("parsing instance: %w", err)
+								}
+								cert, err = api.F3GetCertificate(cctx.Context, instance)
+							} else {
+								cert, err = api.F3GetLatestCertificate(cctx.Context)
+							}
+							if err != nil {
+								return fmt.Errorf("getting finality certificate: %w", err)
+							}
+							if cert == nil {
+								_, _ = fmt.Fprintln(cctx.App.Writer, "No certificate.")
+								return nil
+							}
+
+							return outputFinalityCertificate(cctx, cert)
+						},
+					},
+					{
+						Name: "list",
+						Usage: `Lists a range of F3 finality certificates.
+
+A range may optionally be specified as the first argument to indicate 
+inclusive range of 'from' and 'to' instances in following notation:
+'<from>..<to>'. Either <from> or <to> may be omitted, but not both.
+An omitted instance is interpreted as the latest instance.
+
+If no range is specified, certificates are listed in descending order
+of instance, starting from the latest until the limit is reached, i.e.
+the default range of '..0'.
+
+Examples:
+  * '5..': from instance 5 to the latest instance.
+  * '..0': from the latest instance to instance 0.
+  * '3..2': from instance 3 to instance 2.
+  * '2..3': from instance 2 to instance 3.
+`,
+						ArgsUsage: "[range]",
+						Flags: []cli.Flag{
+							f3FlagOutput,
+							f3FlagInstanceLimit,
+						},
+						Before: func(cctx *cli.Context) error {
+							if count := cctx.NArg(); count > 1 {
+								return fmt.Errorf("too many arguments: expected at most 1 but got %d", count)
+							}
+							return nil
+						},
+						Action: func(cctx *cli.Context) error {
+							api, closer, err := GetFullNodeAPIV1(cctx)
+							if err != nil {
+								return err
+							}
+							defer closer()
+
+							limit := cctx.Int(f3FlagInstanceLimit.Name)
+							fromTo := cctx.Args().First()
+							if fromTo == "" {
+								fromTo = "..0"
+							}
+							r, err := newRanger(fromTo, limit, func() (uint64, error) {
+								latest, err := api.F3GetLatestCertificate(cctx.Context)
+								if err != nil {
+									return 0, fmt.Errorf("getting latest finality certificate: %w", err)
+								}
+								if latest == nil {
+									return 0, errors.New("no latest finality certificate")
+								}
+								return latest.GPBFTInstance, nil
+							})
+							if err != nil {
+								return err
+							}
+
+							var cert *certs.FinalityCertificate
+							for cctx.Context.Err() == nil {
+
+								next, proceed := r.next()
+								if !proceed {
+									return nil
+								}
+								cert, err = api.F3GetCertificate(cctx.Context, next)
+								if err != nil {
+									return fmt.Errorf("getting finality certificate for instance %d: %w", next, err)
+								}
+								if cert == nil {
+									_, _ = fmt.Fprintln(cctx.App.Writer, "No certificate.")
```

Comment:
Also... do we want to print this? E.g., if I ask for `5..` and there are only 3 certs available, getting a "no certificate" line at the end doesn't make a lot of sense.

Reply:
> if I ask for 5.. and there are only 3 certs available, getting a "no certificate" line at the end doesn't make a lot of sense.

In that case, this message will not be printed, because the logic finds the latest first then iterates within that range.

Regardless i think you are right in that it's redundant, because we will error out earlier if there is no latest cert.